internal/isoeditor/rhcos.go

```diff
@@ -326,3 +335,71 @@ func IgnitionImageArchive(ignitionConfig string) ([]byte, error) {
 
 	return compressedBuffer.Bytes(), nil
 }
+
+// GetEmbedArea takes the 24 bytes of a CoreOS ISO and validates it is a
+// CoreOS ISO file before returning the offset into the file (and length) of
+// the embed area.
+func GetEmbedArea(header []byte) (int64, int64, error) {
+	// NOTE: The last 24 bytes should be:
+	// 8 bytes: magic string "coreiso+" tells us this is a CoreOS ISO image
+	// 8 bytes little-endian: offset of embed area from start of ISO file
+	// 8 bytes little-endian: length of embed area
+	// ref: https://github.com/coreos/coreos-installer/blob/80b828caa037bf77968fb472dffe63fe77e15826/src/live.rs#L435
+	if len(header) != 24 {
+		return 0, 0, errors.Errorf("Header unexpected size %v, expected 24", len(header))
+	}
+
+	if !bytes.Equal(header[0:8], []byte(coreosIgnitionHeaderMagic)) {
+		return 0, 0, errors.New("Unrecognized CoreOS ISO image.")
+	}
+
+	offset := int64(binary.LittleEndian.Uint64(header[8:16]))
+	length := int64(binary.LittleEndian.Uint64(header[16:24]))
+	return offset, length, nil
+}
+
+func EmbedIgnition(inputISOPath, outputISOPath, ignitionConfig string) error {
+	coreosIgnitionHeader := make([]byte, coreosIgnitionHeaderSize)
+
+	inputISO, err := os.Open(inputISOPath)
+	if err != nil {
+		return err
+	}
+	defer inputISO.Close()
+
+	if _, err = inputISO.ReadAt(coreosIgnitionHeader, isoSystemAreaSize-coreosIgnitionHeaderSize); err != nil {
+		return err
+	}
+	embedOffset, embedLength, err := GetEmbedArea(coreosIgnitionHeader)
+	if err != nil {
+		return err
+	}
+
+	cpio, err := IgnitionImageArchive(ignitionConfig)
+	if err != nil {
+		return err
+	}
+
+	if len(cpio) > int(embedLength) {
+		return errors.Errorf("Compressed Ignition config is too large: %v > %v", len(cpio), int(embedLength))
+	}
+
+	resultISO, err := os.Create(outputISOPath)
+	if err != nil {
+		return err
+	}
+	defer resultISO.Close()
+
+	if _, err = io.Copy(resultISO, inputISO); err != nil {
+		return err
+	}
+
+	// clear out the embed area
+	embedArea := make([]byte, embedLength)
```

Comment:
I don't think this clear the embed area

Reply:
My understanding was that this would create an array of zeroed bytes with length equal to `embedLength` (https://golang.org/ref/spec#The_zero_value). Is there something I am not seeing?

Reply:
Looks OK to me